lib/pbench/agent/tool_meister.py

```diff
@@ -1518,6 +1539,41 @@ def end_tools(self, data):
                     name,
                 )
                 failures += 1
+
+        # Remove persistent tool temporary working directory
+        directory = data["directory"]
+        tool_dir = self.directories[directory]
+
+        self.logger.debug(
+            "%s: deleting persistent tool tmp directory %s %s",
+            self._hostname,
+            self._group,
+            tool_dir,
+        )
+        unexpected_files = []
+        for dirpath, _, files in os.walk(tool_dir):
+            if files:
+                relative_file_path = {
+                    ",".join(
+                        map(
+                            lambda x: f"{Path(dirpath).relative_to(tool_dir)}/{x}",
+                            files,
+                        )
+                    )
+                }
+                unexpected_files += relative_file_path
+
+        if unexpected_files:
+            self.logger.warning(
+                f"{self._hostname}: unexpected temp files {unexpected_files}"
+            )
```

Comment:
Much better, although I don't get the degenerate dict comprehension for `relative_file_path`. I also think that this list of joined lists representation is a bit weird. Since you're capturing the path of each relative to the common base directory, I would have just extended the `unexpected_files` array and done a single join at the end rather than have a list of comma-lists (assuming there's more than one directory with more than one file each, which doesn't seem to happen in the gold files I examined).

```
         unexpected_files: List[str] = []
￼        for dir, _, files in os.walk(tool_dir):
             dirpath = Path(dir).relative_to(tool_dir)
￼            if files:
￼                relative_file_path = map(
￼                    lambda x: f"{dirpath}/{x}",
￼                    files,
￼                )
￼                unexpected_files += relative_file_path
￼
￼        if unexpected_files:
￼            self.logger.warning(
￼                f"{self._hostname}: unexpected temp files {','.join(unexpected_files)}"
￼            )
```

Reply:
Ah! Thanks I found a bug in my approach, multiple `join` inside a for loop would result in forming a single string in a list, e.g. `["/var/1,/var/2,/var/3"]` instead of `["/var/1","/var/2","/var/3"]` if there are multiple files at the same directory, and we want the latter output.